### ISSUE TITLE
runtime-aware scheduling

### ISSUE THREAD

**Issue by yastij** [kubernetes/kubernetes#72413]
With runtimeClass users now tend to have heterogeneous runtimes across clusters, we might want to make the scheduler runtime-aware to determine if node can host a pod requiring a specific runtime/runtimeClass.

This has the benefit to abstract topology knowledge of the cluster to end-users.

cc @bsalamat @tallclair 

/assign
/sig scheduling
/priority important-soon
/kind feature

**Comment by yastij** [comment]
An end-user that wants to deploy an application might not have the knowledge of cluster’s topology. This will abstract it from him.

**Comment by k82cn** [comment]
> might not have the knowledge of cluster’s topology. 

If user did not know which runtime to use, what can we do?

**Comment by yastij** [comment]
knowing cluster's topology is not the same as knowing which runtime is required for your workload. Users usually know which runtime to use but not on which node it is available.

**Comment by yastij** [comment]
kubelet advertise the runtimeVersion under nodeInfo, but it's not part of the well-known kubernetes labels.
I still expect this work out of the box without having to label hundreds of nodes (I agree that we can make kubelet surface runtime as a label) and start to think about adding affinities.

This is something so common that I'd expect it to be done automatically for the users, as I'm not sure if there's a use case where they'd want being scheduled to a node without the corresponding runtime.

**Comment by Huang-Wei** [comment]
@yastij 's idea (to have a new Predicate supporting this) is based on the assumption that `RuntimeClass` is associated with node spec/status and synced properly. - **(A)**

And @k82cn 's idea is based on the assumption that `RuntimeClass` is associated and synced with node labels properly. - **(B)**

According to [design doc](https://docs.google.com/document/d/1W51yBNTvp0taeEss56GTk8jczqFJ2d6jBeN6sCSlYZU/edit?ts=5bb7d4d6#) and [KEP](https://github.com/kubernetes/enhancements/blob/master/keps/sig-node/0014-runtime-class.md#implementation-history), `RuntimeClass` will be exposed as a new API, and referenced in PodSpec. It only explains in user's perspective, but it doesn't explain how `RuntimeClass` will be associated with node, in which API fashion - via node spec/status, or node labels? @timstclair could you elaborate on this?

**Comment by yastij** [comment]
@Huang-Wei - I'm not sure, but we might find correlation between the runtimeClass of the pod and the runtimeVersion specified in nodeInfo. 
I believe that for now, we do not associate runtimeClass with nodes. The kubelet only reads the runtimeClass to pass runtimeHandler to the underlying CRI.

**Comment by Huang-Wei** [comment]
Right now when you create a fake `RuntimeClass`, no error occurs - neither in validation nor in runtime phase. And I believe the `runtime` info hasn't surfaced in a kube API fashion yet, it's only correlated between CRI runtime and kubelet _internally_:

```console
$ crictl version
Version:  0.1.0
RuntimeName:  docker
RuntimeVersion:  17.03.2-ce
RuntimeApiVersion:  1.27.0
```

So prior to start implementation on scheduler side, `RuntimeName` is supposed to be exposed in node API object - either via spec/status or labels.

**Comment by yastij** [comment]
We only surface the runtimeVersion on the node object under `.spec.nodeInfo.runtimeVersion`, But I agree we might want to just have `RuntimeName` added.

/sig node

**Comment by k82cn** [comment]
>  synced with node labels properly.

Nop, node affinity also support field; for now, we only support "Name" field right now.

My major concern is that, there're several similar "algorithm" there; it increased maintain effort and may impact performance. And if we want to do enhancement, several place need to care about :)

**Comment by Huang-Wei** [comment]
That's true - just like what we did for daemonset pods scheduling.

> My major concern is that, there're several similar "algorithm" there;

Agree. If we can reuse this pattern (`key: <spec field like metadata.name>` in `nodeAffinity...matchFields`), it would save use duplicate efforts. Even if we can reuse this pattern, it's still necessary to add cache internally to index by `spec field` for performance purpose.

**Comment by yastij** [comment]
I’m not against reusing nodeAffinity to implement a runtime predicates, performance-wise it’s still the same as if added by the user and it avoids duplication.

**Comment by yastij** [comment]
cc @resouer

**Comment by tallclair** [comment]
We discussed this a couple months ago, with a few options enumerated in https://docs.google.com/document/d/1W51yBNTvp0taeEss56GTk8jczqFJ2d6jBeN6sCSlYZU/edit?ts=5bb7d4d6

Since writing that, I'm reconsidering built-in scheduler support vs. "injecting" node affinity. The basic idea is the node affinity is declared on the RuntimeClass, and an admission controller updates the PodSpec with the given affinity. I still need to look into how hard it is to reconcile the injected affinity with preexisting (e.g. user-set) rules.

**Comment by k82cn** [comment]
> Since writing that, I'm reconsidering built-in scheduler support vs. "injecting" node affinity. 

emm... Seems I also supported a new native algorithm/policy for runtime-class in this doc :); anyway, I'd like to see whether `NodeAffinity for field` will contribute :)

**Comment by yastij** [comment]
/remove-lifecycle stale

api at #75744

**Comment by yastij** [comment]
/remove-lifecycle stale

@tallclair - once the e2e PR lands we can close this
